File: src/freelm/_ratelimit.py
"""Token bucket for requests-per-minute pacing. Time injected for testability."""
from __future__ import annotations

import threading
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class TokenBucket:
    rate_per_min: float
    capacity: Optional[float] = None
    tokens: Optional[float] = None
    updated: float = 0.0
    _lock: threading.Lock = field(default_factory=threading.Lock, repr=False, compare=False)

    def __post_init__(self) -> None:
        if self.capacity is None:
            self.capacity = max(1.0, float(self.rate_per_min))
        if self.tokens is None:
            self.tokens = self.capacity

    def _refill(self, now: float) -> None:
        if self.updated == 0.0:
            self.updated = now
            return
        dt = now - self.updated
        if dt <= 0:
            return
        self.tokens = min(self.capacity, self.tokens + dt * (self.rate_per_min / 60.0))  # type: ignore[operator]
        self.updated = now

    def peek(self, now: float) -> float:
        with self._lock:
            self._refill(now)
            return self.tokens  # type: ignore[return-value]

    def consume(self, n: float, now: float) -> bool:
        with self._lock:
            self._refill(now)
            if self.tokens >= n:  # type: ignore[operator]
                self.tokens -= n  # type: ignore[operator]
                return True
            return False

    def time_until(self, n: float, now: float) -> float:
        with self._lock:
            self._refill(now)
            if self.tokens >= n:  # type: ignore[operator]
                return 0.0
            deficit = n - self.tokens  # type: ignore[operator]
            return deficit / (self.rate_per_min / 60.0)

File: js/src/ratelimit.ts
/** Token bucket for requests-per-minute pacing. Single-threaded — no lock needed. */
export class TokenBucket {
  capacity: number;
  tokens: number;
  updated = 0;

  constructor(public ratePerMin: number, capacity?: number) {
    this.capacity = capacity ?? Math.max(1, ratePerMin);
    this.tokens = this.capacity;
  }

  private refill(now: number): void {
    if (this.updated === 0) {
      this.updated = now;
      return;
    }
    const dt = now - this.updated;
    if (dt <= 0) return;
    this.tokens = Math.min(this.capacity, this.tokens + dt * (this.ratePerMin / 60));
    this.updated = now;
  }

  peek(now: number): number {
    this.refill(now);
    return this.tokens;
  }

  consume(n: number, now: number): boolean {
    this.refill(now);
    if (this.tokens >= n) {
      this.tokens -= n;
      return true;
    }
    return false;
  }

  timeUntil(n: number, now: number): number {
    this.refill(now);
    if (this.tokens >= n) return 0;
    return (n - this.tokens) / (this.ratePerMin / 60);
  }
}
